### DISCUSSION TITLE
Help freeRASP to improve.  What feature would you love to be included in freeRASP?

### DISCUSSION THREAD

**Discussion by SirionRazzer** [talsec/Free-RASP-Community#32] [Polls]
We have a bunch of great features in mind. Which problems could they help you tackle? Upvote multiple ideas or suggest new ones in the discussion below!

**Comment by SirionRazzer** [comment]
1. **Dynamic TLS/SSL Certificate Pinning**
Prevent network MitM attacks without headaches of statical pinning.

**Comment by SirionRazzer** [comment]
2. **Secure Storage**
Store data securely on the device. Security is enhanced by other RASP features.

**Comment by SirionRazzer** [comment]
3. **VPN Detection**
Enhance security by detecting and preventing certain types of network attacks involving VPNs or proxies.

**Comment by SirionRazzer** [comment]
4. **Screen Cast/Share Detection**
Mitigate certain types of phishing attacks by detecting screen sharing or casting activities.

**Comment by SirionRazzer** [comment]
5. **Keylogger Detection**
Counter 3rd-party malicious keyboard apps that may steal users' credentials.

**Comment by SirionRazzer** [comment]
6. **Location Spoofing Detection**
Stop attacks based on spoofed fake locations by detecting location spoofing attempts.

**Comment by SirionRazzer** [comment]
7. **Multi-instancing Detection**
Identify work profiles, parallel apps, and multiple users to prevent fraudulent registrations on the same device. ([article](https://medium.com/@talsec/fraud-proofing-an-android-app-choosing-the-best-device-id-for-promo-abuse-prevention-aa4a2459637f))

**Comment by SirionRazzer** [comment]
8. **Time Spoofing/Rewinding**
Prevent cheating by detecting attempts to rewind the system time to unlock time-based features in your app.

**Comment by SirionRazzer** [comment]
9. **End-to-end Encryption** - "Evil admin / Man in the middle" protection. Protect data during transmission by ensuring that only the intended parties can access the information. In an end-to-end encrypted communication system, data is encrypted on the sender's device and remains encrypted until it reaches the recipient's device (ie. database)

**Comment by SirionRazzer** [comment]
10. **Enhanced Root Detection** - Identify and deter reverse engineering environments to enhance overall security

**Comment by alexpeju** [comment]
I think it is necessary to improve the obfuscation function, it should be stronger.

**Comment by syakymchuk** [comment]
11. Device Risk Scoring API
